sql/core/src/test/scala/org/apache/spark/sql/SparkSessionExtensionSuite.scala

```diff
@@ -251,6 +276,368 @@ object MyExtensions {
+        val r = rhs.asInstanceOf[ColumnVector]
+        val result = new OnHeapColumnVector(batch.numRows(), dataType)
+        ret = result
+
+        for (i <- 0 until batch.numRows()) {
+          result.appendLong(l + r.getLong(i) + 1) // BUG to show we replaced Add
+        }
+      } else if (lhs.isInstanceOf[ColumnVector]) {
+        val l = lhs.asInstanceOf[ColumnVector]
+        val r = rhs.asInstanceOf[Long]
+        val result = new OnHeapColumnVector(batch.numRows(), dataType)
+        ret = result
+
+        for (i <- 0 until batch.numRows()) {
+          result.appendLong(l.getLong(i) + r + 1) // BUG to show we replaced Add
+        }
+      } else {
+        ret = nullSafeEval(lhs, rhs)
+      }
+    } finally {
+      if (lhs != null && lhs.isInstanceOf[ColumnVector]) {
+        lhs.asInstanceOf[ColumnVector].close()
+      }
+      if (rhs != null && rhs.isInstanceOf[ColumnVector]) {
+        rhs.asInstanceOf[ColumnVector].close()
+      }
+    }
+    ret
+  }
+}
+
+class CannotReplaceException(str: String) extends RuntimeException(str) {
+
+}
+
+case class PreRuleReplaceAddWithBrokenVersion() extends Rule[SparkPlan] {
+  def replaceWithColumnarExpression(exp: Expression): ColumnarExpression = exp match {
+    case a: Alias =>
+      new ColumnarAlias(replaceWithColumnarExpression(a.child),
+        a.name)(a.exprId, a.qualifier, a.explicitMetadata)
+    case att: AttributeReference =>
+      new ColumnarAttributeReference(att.name, att.dataType, att.nullable,
+        att.metadata)(att.exprId, att.qualifier)
+    case lit: Literal =>
+      new ColumnarLiteral(lit.value, lit.dataType)
+    case add: Add if (add.dataType == LongType) &&
+      (add.left.dataType == LongType) &&
+      (add.right.dataType == LongType) =>
+      // Add only supports Longs for now.
+      new BrokenColumnarAdd(replaceWithColumnarExpression(add.left),
+        replaceWithColumnarExpression(add.right))
+    case exp =>
+      throw new CannotReplaceException(s"expression " +
+        s"${exp.getClass} ${exp} is not currently supported.")
+  }
+
+  def replaceWithColumnarPlan(plan: SparkPlan): SparkPlan =
+    try {
+      plan match {
+        case plan: ProjectExec =>
+          new ColumnarProjectExec(plan.projectList.map((exp) =>
+            replaceWithColumnarExpression(exp).asInstanceOf[NamedExpression]),
+            replaceWithColumnarPlan(plan.child))
+        case p =>
+          logWarning(s"Columnar processing for ${p.getClass} is not currently supported.")
+          p.withNewChildren(p.children.map(replaceWithColumnarPlan))
+      }
+    } catch {
+      case exp: CannotReplaceException =>
+        logWarning(s"Columnar processing for ${plan.getClass} is not currently supported" +
+          s"because ${exp.getMessage}")
+        plan
+    }
+
+  override def apply(plan: SparkPlan): SparkPlan = replaceWithColumnarPlan(plan)
+}
+
+case class MyPostRule() extends Rule[SparkPlan] {
+  override def apply(plan: SparkPlan): SparkPlan = plan
```

Comment:
it looks identical plan. We would appreciate it if the testcase would have non-identical plan for ease of review.

Reply:
Sure, I'll look at inserting in a noop of some kind to show that it was replaced.